### PR TITLE
tree: Misc schema factory code cleanups

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

A couple of minor improvements:
1. simplify ArrayToUnion's implementation 
2. document and allow strong typing of TreeNodeSchemaCore.info
3. preserve type of info field for some object schema. This mostly makes debugging and testing typing issues with schema derived types easier, but could be useful for other thing which want field schema access in the future.

## Breaking Changes

Code which makes overly specific assumptions about the typing of schema's .info or the inner workings of ArrayToUnion could break, but this is not expected to impact any reasonable code.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/typed-schema/typeUtils.ts [hunk 1]
```diff
@@ -29,11 +29,7 @@ export function objectToMapTyped<
...
- * Convert a Array type into the type of ReadonlySet.
- *
- * Same as `keyof ListToKeys<T, unknown>` but work for values that are not valid keys.
+ * Convert a Array type into a union of its value types.
  * @public
  */
-export type ArrayToUnion<T extends readonly unknown[]> = T extends readonly (infer TValue)[]
-	? TValue
-	: never;
+export type ArrayToUnion<T extends readonly unknown[]> = T[number];
```
Comment (Josmithr) [inline packages/dds/tree/src/feature-libraries/typed-schema/typeUtils.ts:35]: This doesn't seem right. What's going on here?

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 2]
```diff
@@ -98,7 +98,8 @@ export interface TreeNodeSchemaClass<
 	out TNode = unknown,
 	in TInsertable = never,
 	out ImplicitlyConstructable extends boolean = boolean,
-> extends TreeNodeSchemaCore<Name, Kind, ImplicitlyConstructable> {
+	out Info = unknown,
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:101]: Should we add `@typeParam` docs for this? Or do you think the runtime param documentation is sufficient?

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 3]
```diff
@@ -117,10 +118,16 @@ export interface TreeNodeSchemaCore<
...
 	out ImplicitlyConstructable extends boolean,
+	out Info = unknown,
 > {
 	readonly identifier: Name;
 	readonly kind: Kind;
-	readonly info: unknown;
+
+	/**
+	 * Data used to define this schema.
+	 * The format depends on the kind of node it is for.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:128]: Can we elaborate on this further? Perhaps an `@example` block?

**Review by Josmithr — APPROVED** [review]

File: packages/dds/tree/src/simple-tree/schemaTypes.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:101]: What do you mean by "runtime param documentation is sufficient". This is type parameter to a type. There is no runtime here.

File: packages/dds/tree/src/simple-tree/schemaTypes.ts (hunk 3 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:128]: this is arbitrary data in arbitrary format, selected however the user of this type wants, currently mostly desired for writing compile time assertions for type debugging purposes. I don't think there is much of an example that could be useful here without adding some real odd coupling (like making the example of this depend on implementation choices inside of SchemaFactory.object). I agree could use more clarity though, but a formal example block probably won't work for that,.

File: packages/dds/tree/src/simple-tree/schemaTypes.ts (hunk 2 above)
Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:101]: I should have said property. I'm wondering if there is anything we can say about the `Info` type param that isn't captured by the `info` property.

File: packages/dds/tree/src/simple-tree/schemaTypes.ts (hunk 3 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:130]: Should be better now

File: packages/dds/tree/src/simple-tree/schemaTypes.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:101]: I added a little detail to the field's comment, which I think is enough. No one should really be using this type directly except for the implementation of schemaFactory.

**Pull request merged**